### PR TITLE
Feature/1751 remove deep paging from discovery

### PR DESCRIPTION
**Pull request by anusharanganathan**
This PR is a work in progress and not ready to be reviewed

### CODE REVIEW

**Comment by anusharanganathan** [comment]
* Replace DiscoveryApi search with a scroll for the dump
* Use [query filters](https://github.com/DOAJ/doaj/blob/develop/portality/lib/query_filters.py)
    * Add another query filter to [include source](https://github.com/DOAJ/doaj/blob/develop/portality/api/v1/discovery.py#L362)
    * Replace public_result_filter with a source query filter instead in [settings.py](https://github.com/DOAJ/doaj/blob/develop/portality/settings.py#L342). 
        * The query filter needs to include sealed and ticked from the public_result_filter.
* Scroll gives one record at a time. Just json.dumps each record to file. Keep count and close the file after N number of lines and start a new file.

**Comment by markmacgillivray** [comment]
What does scroll gives one record at a time mean? Can't a scroll return many records?

File: portality/api/v1/discovery.py [hunk 1]
```diff
@@ -121,13 +121,24 @@ def _sanitise(cls, q, page, page_size, sort, search_subs, sort_subs):
...
-
         # calculate the position of the from cursor in the document set
         fro = (page - 1) * page_size
+        # If fro is greater than the max allowed, throw error
+        # using bulk to provide an override when needed
+        max_records = app.config.get("DISCOVERY_MAX_RECORDS_SIZE", 1000)
+        if not bulk and fro >= max_records:
+            message = """Looks like you want to view results beyond {x}. \
+If you would like to see more results, you can download all of our results \
+from link""".format(x=max_records)
```
Comment (Steven-Eardley) [inline portality/api/v1/discovery.py]: This is a placeholder for the full data dump URL, but this should be either be meaningful without that or we shouldn't release this until the dump is ready.

File: portality/api/v1/discovery.py [hunk 2]
```diff
@@ -221,25 +231,39 @@ def _make_response(cls, endpoint, res, q, page, page_size, sort,
             "results" : obs
         }
 
+        # FIXME: URLs constructed without helper if exception is thrown
+        #       as application context is not available
+        s = sort
+        if s == None:
```
Comment (Steven-Eardley) [inline portality/api/v1/discovery.py]:
```suggestion
        if s is None:
```

File: portality/api/v1/discovery.py [hunk 3]
```diff
@@ -221,25 +231,39 @@ def _make_response(cls, endpoint, res, q, page, page_size, sort,
             "results" : obs
         }
 
+        # FIXME: URLs constructed without helper if exception is thrown
```
Comment (Steven-Eardley) [inline portality/api/v1/discovery.py]: Can you explain in which circumstance we don't have an application context?

File: portality/api/v1/discovery.py [hunk 4]
```diff
@@ -221,25 +231,39 @@ def _make_response(cls, endpoint, res, q, page, page_size, sort,
...
+        #       as application context is not available
+        s = sort
+        if s == None:
+            s = ''
         if previous_page is not None:
-            result["prev"] = app.config['BASE_URL'] + url_for(app.config['API_BLUEPRINT_NAME'] + '.' + endpoint, search_query=q, page=previous_page, pageSize=page_size, sort=sort)
+            try:
+                result["prev"] = app.config['BASE_URL'] + url_for(app.config['API_BLUEPRINT_NAME'] + '.' + endpoint, search_query=q, page=previous_page, pageSize=page_size, sort=sort)
+            except:
```
Comment (Steven-Eardley) [inline portality/api/v1/discovery.py]: too broad an exception - when and why does the line above fail?

File: portality/scripts/api_export.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+def _save_file(typ):
+    # Create a dir for today and save all files in there
+    data = json.dumps(results, cls=ModelJsonEncoder)
+    filename = os.path.join(today, "{typ}_{page}.json".format(typ=typ, page=page))
+    output_file = tmpStore.path(container, filename, create_container=True, must_exist=False)
+    dir = os.path.dirname(output_file)
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+    print("Saving file {filename}".format(filename=filename))
+    out = codecs.open(output_file, "wb", "utf-8")
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: I prefer the ```with``` syntax here

File: portality/scripts/api_export.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    out.write(data)
+    out.close
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: if not above, this should be `out.close()`

File: portality/scripts/api_export.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    return
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: No need for `return` statement; it's implied at the end of a function

File: portality/scripts/api_export.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def _get_last_page():
+    url = results.data.get('last', '')
+    last_page = None
+    if url:
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: no case for `not url` - is that correct?

Comment (Steven-Eardley) [reply portality/scripts/api_export.py]: You can remove the `else` so the `return none` covers both cases (if that's intended)

File: portality/scripts/api_export.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            total_size += os.path.getsize(fp)
+    return total_size
+
+
+def _prune_container():
+    # Delete all files and dirs in the container that does not contain today's date
+    for dir in mainStore.list(container):
+        if today not in dir:
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]:
```suggestion
        if dates.today() not in dir:
```

File: portality/scripts/api_export.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            mainStore.delete(target_name=dir)
+
+
+def _copy_on_complete(path):
+    name = os.path.basename(path)
+    raw_size = _get_dir_size(path)
+    print("Compressing temporary file {x}".format(x=path, y=raw_size))
+    zipped_name = name + ".gz"
+    dir = os.path.dirname(path)
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: this overwrites python's `dir`

File: portality/scripts/api_export.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    if url:
+        last_page = parse_qs(urlparse(url).query).get('page', [])
+    if last_page:
+        return int(last_page[0])
+    else:
+        return None
+
+def _get_dir_size(path):
+    total_size = 0
+    for dirpath, dirnames, filenames in os.walk(start_path):
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]:
```suggestion
    for dirpath, dirnames, filenames in os.walk(path):
```
no variable `start_path`

File: portality/scripts/api_export.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    # Delete all files and dirs in the container that does not contain today's date
+    for dir in mainStore.list(container):
+        if today not in dir:
+            mainStore.delete(target_name=dir)
+
+
+def _copy_on_complete(path):
+    name = os.path.basename(path)
+    raw_size = _get_dir_size(path)
+    print("Compressing temporary file {x}".format(x=path, y=raw_size))
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: you're not using value `y` that the poor little computer worked so hard to calculate.
Also, if you're not using meaningful formatting names you may as well use `"{0} and {1}".format(positional, args)`

File: portality/scripts/api_export.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+    tmpStore = StoreFactory.tmp()
+    mainStore = StoreFactory.get("api_data")
+    container = app.config.get("STORE_ANON_DATA_CONTAINER")
+
+    if args.clean:
+        mainStore.delete(container)
+
+    # create dir with today's date
+    today = datetime.now().strftime("%Y%m%d")
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: we already have his function, `portality.lib.dates.today()` (as above in line 47)
I think the format with dashes should still play nicely with s3, right?

File: portality/scripts/api_export.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+if __name__ == '__main__':
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-t", "--type", choices=['articles','journals', 'all'], help="type of data to export. ")
+    parser.add_argument("-c", "--clean", action="store_false", help="Clean any pre-existing output before continuing")
+    parser.add_argument("-p", "--prune", action="store_true", help="Delete previous backups if anyafter running current backup")
+    args = parser.parse_args()
+
+    tmpStore = StoreFactory.tmp()
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: I'm not a fan of global variables in Python

File: portality/scripts/api_export.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+        page = 1
+        last_page = 1
+        while page <= last_page:
+            print("journals page {p} of {l}".format(p=page, l=last_page))
+            try:
+                results = DiscoveryApi.search_journals(query, page, page_size, sort, bulk=True)
+            except DiscoveryException as e:
+                raise Api400Error(e.message)
+            _save_file(typ='journals')
+            new_last_page = _get_last_page()
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: as my comment regarding globals above, I'd much prefer `_get_last_page()` to take `results` as an argument

File: portality/scripts/api_export.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+            if new_last_page is not None:
+                last_page = new_last_page
+            page = page + 1
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]:
```suggestion
            page += 1
```

File: portality/scripts/api_export.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    zipped_path = os.path.join(dir, zipped_name)
+    tar = tarfile.open(zipped_path, "w:gz")
+    tar.add(path)
+    tar.close()
+    zipped_size = os.path.getsize(zipped_path)
+    print("Storing from temporary file {x} ({y} bytes)".format(x=zipped_name, y=zipped_size))
+    mainStore.store(container, zipped_name, source_path=zipped_path)
+    tmpStore.delete(container, name)
+    tmpStore.delete(container, zipped_name)
+    return
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: another unnecessary return statement

File: portality/scripts/api_export.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+            new_last_page = _get_last_page()
+            if new_last_page is not None:
+                last_page = new_last_page
+            page = page + 1
+
+    print(dates.now() + ": done\n")
+
+    # Copy the source directory to main store
+    out_dir = tmpStore.path(container, today)
+    _copy_on_complete(out_dir)
```
Comment (Steven-Eardley) [inline portality/scripts/api_export.py]: do we need a `try:` here? It'd be a shame to discard our work from the temp store if the compression or save fails in `_copy_on_complete()`

**Comment by anusharanganathan** [comment]
@Steven-Eardley this PR is a work in progress. I guess I should have mentioned it. I raised the PR so Richard could easily review what I had done. We did this on late Wednesday evening. I haven't yet gone through all of your comments but will do so as I am doing the work

**Comment by anusharanganathan** [comment]
I have split this issue into 2 parts split, so it's easier to manage the coding and review process. Also, most of the code here will move to the feature relating to removing deep paging. See [#1750](https://github.com/DOAJ/doajPM/issues/1750) and [#1751](https://github.com/DOAJ/doajPM/issues/1751)

**Pull request closed without merging**